src/icache.rs

```diff
@@ -0,0 +1,200 @@
+//! Instruction Cache management
+//!
+//! To enable the the ICACHE:
+//!
+//! ```
+//! let dp = ...;            // Device peripherals
+//! let mut icache = dp.ICACHE.constrain();
+//!
+//! icache.enable();
+//! ```
+//!
+//! To invalidate the cache, use the ICache::invalidate() method:
+//!
+//! ```
+//! icache.invalidate();
+//! ```
+//!
+//! Performance monitoring of the cache is possible using the cache hit and
+//! miss counters:
+//!
+//! ```
+//! icache.enable_hit_counter();
+//! icache.enable_miss_counter();
+//!
+//! // do something
+//!
+//! let hit_count = icache.hit_count();
+//! let miss_count = icache.miss_count();
+//! ```
+
+use crate::stm32::ICACHE;
+
+pub trait ICacheExt {
+    fn constrain(self) -> ICache;
+}
+
+impl ICacheExt for ICACHE {
+    fn constrain(self) -> ICache {
+        ICache::new(self)
+    }
+}
+
+pub struct ICache {
+    icache: ICACHE,
+}
+
+impl ICache {
+    fn new(icache: ICACHE) -> Self {
+        Self { icache }
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        cortex_m::asm::dsb();
+        cortex_m::asm::isb();
+
+        self.icache.cr().read().en().bit_is_set()
+    }
+
+    fn wait_for_busy_complete(&self) {
+        while self.icache.sr().read().busyf().bit_is_set() {}
+    }
+
+    /// Enable ICACHE in default operating mode (N-way associative)
+    pub fn enable(&mut self) {
+        self.enable_n_way()
+    }
+
+    /// Enable the ICACHE in N-way associative mode
+    pub fn enable_n_way(&mut self) {
+        if self.is_enabled() {
+            return;
```

Comment:
Shouldn't we return an error so the user knows to do something different instead of silently ignoring their request? Same for all the modes below. Even returning `Err(())` might be enough.

Reply:
Good point about ignoring the request. I'll have a look at the peripheral again (I wrote this a while ago) and see what would be appropriate. I'm thinking that maybe switching gracefully between N-way and direct mapped modes might be better for this and the `enable_direct_mapped` mode. I think for `disable` simply returning if it is already disabled seems fine: it might be called more than once, and that's fine I think?

For the reset functions on the counters, I think the counter is reset if the peripheral is disabled so that's probably fine, but I'll double check. The asserts on enabling the hit counter are more aggressive than other calls where things are ignored, so I can take a look at those.